drivers/spi/spi_sedi.c

```diff
@@ -0,0 +1,435 @@
+#include "spi_context.h"
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/pm/device.h>
+
+#define SPI_NOT_USE_DMA (-1)
+#define SPI_LINE_MODE_OFF (16)
+#define SPI_DMA_THRESHOLD (32)
+
+struct spi_sedi_config {
+	sedi_spi_t spi_device;
+	void (*irq_config)(void);
+};
+
+struct spi_sedi_data {
+	struct spi_context ctx;
+	bool tx_data_updated;
+	bool rx_data_updated;
+	uint32_t tx_dummy_len;
+	uint32_t rx_dummy_len;
+	bool is_locked;
+};
+
+static int spi_sedi_configure(const struct device *dev,
+			      const struct spi_config *config)
+{
+	struct spi_sedi_data *data = dev->data;
+	const struct spi_sedi_config *info = dev->config;
+	uint32_t word_size, cpol, cpha, loopback;
+
+	if (spi_context_configured(&data->ctx, config) == true) {
+		return 0;
+	}
+
+	word_size = SPI_WORD_SIZE_GET(config->operation);
+	sedi_spi_control(info->spi_device, SEDI_SPI_IOCTL_DATA_WIDTH,
+			 word_size);
+
+	/* CPOL and CPHA */
+	cpol = SPI_MODE_GET(config->operation) & SPI_MODE_CPOL;
+	cpha = SPI_MODE_GET(config->operation) & SPI_MODE_CPHA;
+
+	if ((cpol == 0) && (cpha == 0)) {
+		sedi_spi_control(info->spi_device, SEDI_SPI_IOCTL_CPOL0_CPHA0,
+				 0);
+	} else if ((cpol == 0) && (cpha == 1U)) {
+		sedi_spi_control(info->spi_device, SEDI_SPI_IOCTL_CPOL0_CPHA1,
+				 0);
+	} else if ((cpol == 1) && (cpha == 0U)) {
+		sedi_spi_control(info->spi_device, SEDI_SPI_IOCTL_CPOL1_CPHA0,
+				 0);
+	} else {
+		sedi_spi_control(info->spi_device, SEDI_SPI_IOCTL_CPOL1_CPHA1,
+				 0);
+	}
+
+	/* MSB and LSB */
+	if (config->operation & SPI_TRANSFER_LSB) {
+		sedi_spi_control(info->spi_device, SEDI_SPI_IOCTL_LSB, 0);
+	}
+
+	/* Set loopack */
+	loopback = SPI_MODE_GET(config->operation) & SPI_MODE_LOOP;
+	sedi_spi_control(info->spi_device, SEDI_SPI_IOCTL_LOOPBACK, loopback);
+
+	/* Set baudrate */
+	sedi_spi_control(info->spi_device, SEDI_SPI_IOCTL_SPEED_SET,
+			 config->frequency);
+
+	sedi_spi_control(info->spi_device, SEDI_SPI_IOCTL_CS_HW, config->slave);
+
+	/* Set LOCK_ON */
+	if ((config->operation) & SPI_LOCK_ON) {
+		data->is_locked = true;
+	}
+
+	data->ctx.config = config;
+	spi_context_cs_control(&data->ctx, true);
+
+	return 0;
+}
+
+static int transceive(const struct device *dev, const struct spi_config *config,
+		      const struct spi_buf_set *tx_bufs,
+		      const struct spi_buf_set *rx_bufs, bool asynchronous,
+		      spi_callback_t cb,
+		      void *userdata)
+{
+	const struct spi_sedi_config *info = dev->config;
+	struct spi_sedi_data *spi = dev->data;
+	struct spi_context *ctx = &spi->ctx;
+	int ret;
+	uint32_t transfer_bytes = 0;
+	uint8_t *data_out = NULL, *data_in = NULL;
+	uint32_t i, dummy_len = 0;
+	const struct spi_buf *buf;
+	bool is_multibufs = false;
+
+	if (config != NULL) {
+		spi_context_lock(&spi->ctx, asynchronous, cb, userdata, config);
+	}
+	pm_device_busy_set(dev);
+
+	/* If device locked, will return error, need to release first */
+	if ((spi->is_locked) && (!spi_context_configured(&spi->ctx, config))) {
```

Comment:
please revise all is_locked logics.
it seems the device locking SPI_LOCK_ON is totally handled by spi_context.h now.

Reply:
Will do.